File: module5_explainability/build_summary_figure.m
function [fig_handle, summary_canvas_img] = build_summary_figure(screening_data, xai_maps, clinical_text, save_path)
% BUILD_SUMMARY_FIGURE Creates a 2x3 multi-panel diagnostic summary canvas.
%
% Inputs:
%   screening_data    - Struct with images, metadata, dr_grading_result
%   xai_maps          - Struct with gradcam_overlay, lesion_overlay
%   clinical_text     - Struct with diagnostic_summary, recommendations
%   save_path         - (Optional) Filepath to export PNG image
%
% Outputs:
%   fig_handle        - Figure handle
%   summary_canvas_img - RGB matrix of the combined 2x3 grid canvas

    if nargin < 4; save_path = ''; end

    % Create invisible figure for headless / automated execution
    fig_handle = figure('Visible', 'off', 'Position', [100, 100, 1200, 800], 'Color', [0.96, 0.96, 0.98]);

    % Panel 1: Raw Fundus Image
    subplot(2, 3, 1);
    imshow(screening_data.raw_image);
    title('1. Raw Fundus Image', 'FontSize', 12, 'FontWeight', 'bold');
    
    % Panel 2: Enhanced Fundus Image
    subplot(2, 3, 2);
    imshow(screening_data.enhanced_image);
    title('2. Enhanced Image (Module 2)', 'FontSize', 12, 'FontWeight', 'bold');
    
    % Panel 3: Lesion Segmentation Overlay
    subplot(2, 3, 3);
    imshow(xai_maps.lesion_overlay);
    title('3. Lesion Segmentation (Module 3)', 'FontSize', 12, 'FontWeight', 'bold');
    
    % Panel 4: Grad-CAM XAI Heatmap
    subplot(2, 3, 4);
    imshow(xai_maps.gradcam_overlay);
    title('4. Grad-CAM XAI Heatmap (Module 5)', 'FontSize', 12, 'FontWeight', 'bold');
    
    % Panel 5: Class Probabilities Bar Chart
    subplot(2, 3, 5);
    probs = screening_data.dr_grading_result.class_probabilities;
    b = bar(0:4, probs * 100, 'FaceColor', [0.2, 0.5, 0.8]);
    ylim([0, 100]);
    xlabel('DR Grade (0:Normal ... 4:PDR)', 'FontSize', 10);
    ylabel('Probability (%)', 'FontSize', 10);
    title('5. DR Grade Probabilities', 'FontSize', 12, 'FontWeight', 'bold');
    grid on;
    
    % Highlight predicted grade bar
    pred_grade = screening_data.dr_grading_result.predicted_grade;
    hold on;
    bar(pred_grade, probs(pred_grade + 1) * 100, 'FaceColor', [0.85, 0.25, 0.2]);
    hold off;
    
    % Panel 6: Textual Clinical Summary Box
    subplot(2, 3, 6);
    axis off;
    
    summary_box_text = sprintf([ ...
        'PATIENT METADATA:\n' ...
        '  ID: %s  |  Eye: %s\n' ...
        '  Name: %s (Age: %d)\n\n' ...
        'DIAGNOSTIC RESULT:\n' ...
        '  IQA Status: %s (Score: %.2f)\n' ...
        '  Grade: %s\n' ...
        '  Confidence: %.1f%%\n\n' ...
        'RECOMMENDATION:\n' ...
        '  %s' ...
    ], ...
    screening_data.patient_id, screening_data.eye_side, ...
    screening_data.patient_name, screening_data.age, ...
    screening_data.iqa_result.status, screening_data.iqa_result.quality_score, ...
    screening_data.dr_grading_result.grade_label, ...
    screening_data.dr_grading_result.confidence * 100, ...
    clinical_text.recommendations);

    text(0.05, 0.95, summary_box_text, ...
        'Units', 'normalized', ...
        'VerticalAlignment', 'top', ...
        'FontSize', 9.5, ...
        'FontName', 'Courier', ...
        'BackgroundColor', [1, 1, 1], ...
        'EdgeColor', [0.7, 0.7, 0.8], ...
        'Margin', 8);
    title('6. Clinical Summary', 'FontSize', 12, 'FontWeight', 'bold');

    % Capture canvas frame as image matrix
    drawnow;
    frame = getframe(fig_handle);
    summary_canvas_img = frame.cdata;

    % Export to PNG if path provided
    if ~isempty(save_path)
        if exist('exportgraphics', 'file')
            exportgraphics(fig_handle, save_path, 'Resolution', 150);
        else
            imwrite(summary_canvas_img, save_path);
        end
    end
end

File: module5_explainability python/module5_explainability python/module5_explainability/src/build_summary_figure.py
import matplotlib.pyplot as plt
import numpy as np
import io
from PIL import Image

def build_summary_figure(screening_data, xai_maps, clinical_text, save_path=None):
    """
    Creates a 2x3 multi-panel diagnostic summary canvas.
    """
    fig, axes = plt.subplots(2, 3, figsize=(15, 10), facecolor='#F5F5F9')
    axes = axes.flatten()
    
    # Pre-configure all axes
    for ax in axes:
        ax.axis('off')
        
    # Panel 1: Raw Fundus Image
    axes[0].imshow(screening_data['raw_image'][:, :, ::-1] if len(screening_data['raw_image'].shape) == 3 else screening_data['raw_image'])
    axes[0].set_title('1. Raw Fundus Image', fontsize=12, fontweight='bold')
    
    # Panel 2: Enhanced Fundus Image
    enhanced_img = screening_data.get('enhanced_image', screening_data['raw_image'])
    axes[1].imshow(enhanced_img[:, :, ::-1] if len(enhanced_img.shape) == 3 else enhanced_img)
    axes[1].set_title('2. Enhanced Image (Module 2)', fontsize=12, fontweight='bold')
    
    # Panel 3: Lesion Segmentation Overlay
    axes[2].imshow(xai_maps['lesion_overlay'][:, :, ::-1] if len(xai_maps['lesion_overlay'].shape) == 3 else xai_maps['lesion_overlay'])
    axes[2].set_title('3. Lesion Segmentation (Module 3)', fontsize=12, fontweight='bold')
    
    # Panel 4: Grad-CAM XAI Heatmap
    axes[3].imshow(xai_maps['gradcam_overlay'][:, :, ::-1] if len(xai_maps['gradcam_overlay'].shape) == 3 else xai_maps['gradcam_overlay'])
    axes[3].set_title('4. Grad-CAM XAI Heatmap (Module 5)', fontsize=12, fontweight='bold')
    
    # Panel 5: Class Probabilities Bar Chart
    ax5 = axes[4]
    ax5.axis('on')
    probs = np.array(screening_data['dr_grading_result']['class_probabilities']) * 100
    bars = ax5.bar(range(5), probs, color='#337FB2')
    ax5.set_ylim([0, 100])
    ax5.set_xlabel('DR Grade (0:Normal ... 4:PDR)', fontsize=10)
    ax5.set_ylabel('Probability (%)', fontsize=10)
    ax5.set_title('5. DR Grade Probabilities', fontsize=12, fontweight='bold')
    ax5.grid(True, linestyle='--', alpha=0.6)
    
    pred_grade = screening_data['dr_grading_result']['predicted_grade']
    bars[pred_grade].set_color('#D94033') # Highlight predicted grade
    
    # Panel 6: Textual Clinical Summary Box
    ax6 = axes[5]
    summary_box_text = (
        "PATIENT METADATA:\n"
        f"  ID: {screening_data.get('patient_id', 'N/A')}  |  Eye: {screening_data.get('eye_side', 'N/A')}\n"
        f"  Name: {screening_data.get('patient_name', 'N/A')} (Age: {screening_data.get('age', 'N/A')})\n\n"
        "DIAGNOSTIC RESULT:\n"
        f"  IQA Status: {screening_data.get('iqa_result', {}).get('status', 'N/A')} "
        f"(Score: {screening_data.get('iqa_result', {}).get('quality_score', 0.0):.2f})\n"
        f"  Grade: {screening_data.get('dr_grading_result', {}).get('grade_label', 'N/A')}\n"
        f"  Confidence: {screening_data.get('dr_grading_result', {}).get('confidence', 0.0)*100:.1f}%\n\n"
        "RECOMMENDATION:\n"
        f"  {clinical_text.get('recommendations', 'N/A')}"
    )
    
    ax6.text(0.05, 0.95, summary_box_text,
             transform=ax6.transAxes,
             verticalalignment='top',
             fontsize=9.5,
             family='monospace',
             bbox=dict(facecolor='white', edgecolor='#B3B3CC', boxstyle='round,pad=0.8', alpha=0.9))
    ax6.set_title('6. Clinical Summary', fontsize=12, fontweight='bold')
    
    plt.tight_layout()
    
    # Render canvas to numpy array
    fig.canvas.draw()
    img_data = np.asarray(fig.canvas.buffer_rgba())
    summary_canvas_img = img_data[:, :, :3] # Convert RGBA to RGB
    
    if save_path:
        plt.savefig(save_path, dpi=150, bbox_inches='tight')
        
    plt.close(fig)
    return fig, summary_canvas_img
